src/processor/intents/spawns/_charge-energy.js

```diff
@@ -4,10 +4,28 @@ var _ = require('lodash'),
     C = driver.constants;
 
 
-module.exports = function(spawn, roomObjects, cost, bulk, roomController) {
-
-    var spawns = _.filter(roomObjects, i => i.type == 'spawn' && i.user == spawn.user && !i.off);
-    var extensions = _.filter(roomObjects, i => i.type == 'extension' && i.user == spawn.user && !i.off);
+module.exports = function(spawn, roomObjects, cost, bulk, energyStructures) {
+
+    var spawns = [];
+    var extensions = [];
+
+    if(energyStructures) {
+        _.forEach(energyStructures, id => {
+            let energyStructure = roomObjects[id];
+            if(!energyStructure || energyStructure.off || energyStructure.user !== spawn.user) {
+                return;
+            }
+
+            if(energyStructure.type === 'spawn'){
+                spawns.push(energyStructure);
+            } else if(energyStructure.type === 'extension'){
+                extensions.push(energyStructure);
+            }
+        });
+    } else {
+        spawns = _.filter(roomObjects, i => i.type == 'spawn' && i.user == spawn.user && !i.off);
```

Comment:
For the single list we can use `( i.type == 'spawn' || i.type == 'extension')` saves CPU for filtering as well as we only loop all buildings once.

Reply:
Agreed. I wouldn't mind optimizing this call. Iterating through roomObjects can be quite expensive.

However, my goal was to minimize the footprint of my changes. This was the current implementation. If @artch agrees, I can adjust this to a single iteration.

If you two are interested, I can also swap to using a hardcore for loop instead of _.forEach/_.filter. It will be less readable, but a small increase in performance.